File: programmers/lv1/가장 많이 받은 선물/가장 많이 받은 선물.js
// https://school.programmers.co.kr/learn/courses/30/lessons/258712

function solution(friends, gifts) {
  let result = Array(friends.length).fill(0);
  let giftTable = Array.from({ length: friends.length }, () => [[], []]);

  // giftTable 알맞게 초기화 작업
  for (const gift of gifts) {
    const [a, b] = gift.split(" ");
    // 선물을 준 친구 index
    const aidx = friends.indexOf(a);
    // 선물을 받은 친구 index
    const bidx = friends.indexOf(b);

    giftTable[aidx][0].push(b);
    giftTable[bidx][1].push(a);
  }

  for (let aidx = 0; aidx < friends.length; aidx++) {
    for (let bidx = aidx; bidx < friends.length; bidx++) {
      if (aidx === bidx) {
        continue;
      }
      const avalue = friends[aidx];
      const bvalue = friends[bidx];
      // a 가 b에게 준 선물 개수
      const aToBPresent = giftTable[aidx][0].filter(
        (value) => value === bvalue
      ).length;
      // b 가 a에게 준 선물 개수
      const bToAPresent = giftTable[bidx][0].filter(
        (value) => value === avalue
      ).length;

      if (aToBPresent < bToAPresent) {
        result[bidx] += 1;
      } else if (bToAPresent < aToBPresent) {
        result[aidx] += 1;
      } else {
        // 선물 지수 비교
        const aPresentNum =
          giftTable[aidx][0].length - giftTable[aidx][1].length;
        const bPresentNum =
          giftTable[bidx][0].length - giftTable[bidx][1].length;

        if (aPresentNum < bPresentNum) {
          result[bidx] += 1;
        } else if (bPresentNum < aPresentNum) {
          result[aidx] += 1;
        }
      }
    }
  }

  return Math.max(...result);
}

File: programmers/lv1/가장 많이 받은 선물/가장 많이 받은 선물.py
# https://school.programmers.co.kr/learn/courses/30/lessons/258712

def solution(friends, gifts):
  giftTable = [[[] for _ in range(2)] for _ in range(len(friends))]
  result = [0 for _ in range(len(friends))]

  for gift in gifts:
    a, b = gift.split(" ")
    aidx= friends.index(a)
    giftTable[aidx][0].append(b)

    bidx = friends.index(b)
    giftTable[bidx][1].append(a)
  
  for aidx,avalue in enumerate(friends):
    for bidx in range(aidx, len(friends)):
      bvalue = friends[bidx]
      if aidx == bidx:
        continue
      aForbGift = giftTable[aidx][0].count(bvalue)
      bForaGift = giftTable[bidx][0].count(avalue)

      if (aForbGift < bForaGift):
        result[bidx]+=1
      elif (bForaGift < aForbGift):
        result[aidx]+=1
      else:
        aPresentNum = len(giftTable[aidx][0]) - len(giftTable[aidx][1])
        bPresentNum = len(giftTable[bidx][0]) - len(giftTable[bidx][1])
        if (aPresentNum < bPresentNum):
          result[bidx]+=1
        elif (bPresentNum < aPresentNum):
          result[aidx]+=1

  return max(result)


# 좋은 풀이
def goodSolution(friends, gifts):
    # 친구 이름과 index 값 1대1 매칭
    f = {v: i for i, v in enumerate(friends)}
    # 친구의 수
    l = len(friends)
    # 선물 지수
    p = [0] * l
    answer = [0] * l
    gr = [[0] * l for _ in range(l)]
    for i in gifts:
        a, b = i.split()
        gr[f[a]][f[b]] += 1
    for i in range(l):
        p[i] = sum(gr[i]) - sum([k[i] for k in gr])

    for i in range(l):
        for j in range(l):
            if gr[i][j] > gr[j][i]:
                answer[i] += 1
            elif gr[i][j] == gr[j][i]:
                if p[i] > p[j]:
                    answer[i] += 1
    return max(answer)


